File: app/src/main/java/dev/lumamd/viewer/core/FootnoteProcessor.java
package dev.lumamd.viewer.core;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class FootnoteProcessor {
    private static final Pattern DEFINITION = Pattern.compile(
            "^\\s{0,3}\\[\\^([^]]+)]\\s*:\\s*(.*)$");
    private static final char TOKEN = '\u0005';

    private FootnoteProcessor() {
    }

    static Result extract(String source) {
        String[] lines = source.split("\\n", -1);
        Map<String, String> definitions = collectDefinitions(lines);
        Map<String, Integer> numbers = new LinkedHashMap<String, Integer>();
        Map<String, Integer> occurrences = new LinkedHashMap<String, Integer>();
        List<Reference> references = new ArrayList<Reference>();

        boolean inFence = false;
        for (int index = 0; index < lines.length; index++) {
            String trimmed = lines[index].trim();
            if (isFence(trimmed)) {
                inFence = !inFence;
                continue;
            }
            if (!inFence && !DEFINITION.matcher(lines[index]).matches()) {
                lines[index] = replaceReferences(
                        lines[index],
                        definitions,
                        numbers,
                        occurrences,
                        references);
            } else if (!inFence) {
                lines[index] = "";
            }
        }
        return new Result(join(lines), definitions, numbers, occurrences, references);
    }

    private static Map<String, String> collectDefinitions(String[] lines) {
        Map<String, String> definitions = new LinkedHashMap<String, String>();
        boolean inFence = false;
        for (String line : lines) {
            String trimmed = line.trim();
            if (isFence(trimmed)) {
                inFence = !inFence;
                continue;
            }
            if (inFence) {
                continue;
            }
            Matcher definition = DEFINITION.matcher(line);
            if (definition.matches() && !definitions.containsKey(definition.group(1))) {
                definitions.put(definition.group(1), definition.group(2));
            }
        }
        return definitions;
    }

    private static String replaceReferences(
            String line,
            Map<String, String> definitions,
            Map<String, Integer> numbers,
            Map<String, Integer> occurrences,
            List<Reference> references) {
        StringBuilder output = new StringBuilder(line.length());
        boolean inCode = false;
        int cursor = 0;
        while (cursor < line.length()) {
            char current = line.charAt(cursor);
            if (current == '`') {
                inCode = !inCode;
                output.append(current);
                cursor++;
                continue;
            }
            if (!inCode && current == '[' && cursor + 3 < line.length()
                    && line.charAt(cursor + 1) == '^') {
                int end = line.indexOf(']', cursor + 2);
                if (end > cursor + 2) {
                    String key = line.substring(cursor + 2, end);
                    if (definitions.containsKey(key)) {
                        Integer number = numbers.get(key);
                        if (number == null) {
                            number = numbers.size() + 1;
                            numbers.put(key, number);
                        }
                        int occurrence = occurrences.containsKey(key)
                                ? occurrences.get(key) + 1
                                : 1;
                        occurrences.put(key, occurrence);
                        Reference reference = new Reference(
                                key,
                                number,
                                occurrence,
                                references.size());
                        references.add(reference);
                        output.append(TOKEN).append(reference.tokenIndex).append(TOKEN);
                        cursor = end + 1;
                        continue;
                    }
                }
            }
            output.append(current);
            cursor++;
        }
        return output.toString();
    }

    private static boolean isFence(String line) {
        return line.startsWith("```") || line.startsWith("~~~");
    }

    private static String join(String[] lines) {
        StringBuilder output = new StringBuilder();
        for (int index = 0; index < lines.length; index++) {
            if (index > 0) {
                output.append('\n');
            }
            output.append(lines[index]);
        }
        return output.toString();
    }

    private static String safeId(String key) {
        String value = key.toLowerCase(Locale.ROOT)
                .replaceAll("[^\\p{L}\\p{N}_-]+", "-")
                .replaceAll("(^-|-$)", "");
        return value.isEmpty() ? "note" : value;
    }

    static final class Result {
        private final String source;
        private final Map<String, String> definitions;
        private final Map<String, Integer> numbers;
        private final Map<String, Integer> occurrences;
        private final List<Reference> references;

        private Result(
                String source,
                Map<String, String> definitions,
                Map<String, Integer> numbers,
                Map<String, Integer> occurrences,
                List<Reference> references) {
            this.source = source;
            this.definitions = definitions;
            this.numbers = numbers;
            this.occurrences = occurrences;
            this.references = references;
        }

        String getSource() {
            return source;
        }

        String finish(String documentHtml) {
            String html = documentHtml;
            for (Reference reference : references) {
                html = html.replace(
                        TOKEN + String.valueOf(reference.tokenIndex) + TOKEN,
                        referenceHtml(reference));
            }
            return html + sectionHtml();
        }

        private String referenceHtml(Reference reference) {
            String id = safeId(reference.key);
            String referenceId = "fnref-" + id
                    + (reference.occurrence > 1 ? "-" + reference.occurrence : "");
            return "<sup class=\"footnote-ref\"><a id=\"" + InlineMarkdown.escape(referenceId)
                    + "\" href=\"#fn-" + InlineMarkdown.escape(id)
                    + "\" aria-label=\"Footnote " + reference.number + "\">"
                    + reference.number + "</a></sup>";
        }

        private String sectionHtml() {
            if (numbers.isEmpty()) {
                return "";
            }
            StringBuilder html = new StringBuilder()
                    .append("<section class=\"footnotes\" aria-label=\"Footnotes\">")
                    .append("<hr><ol>");
            for (Map.Entry<String, Integer> numbered : numbers.entrySet()) {
                String key = numbered.getKey();
                String id = safeId(key);
                html.append("<li id=\"fn-").append(InlineMarkdown.escape(id)).append("\">")
                        .append("<span class=\"footnote-content\">")
                        .append(InlineMarkdown.render(definitions.get(key)))
                        .append("</span><span class=\"footnote-backlinks\">");
                int count = occurrences.get(key);
                for (int occurrence = 1; occurrence <= count; occurrence++) {
                    String referenceId = "fnref-" + id
                            + (occurrence > 1 ? "-" + occurrence : "");
                    html.append("<a class=\"footnote-backref\" href=\"#")
                            .append(InlineMarkdown.escape(referenceId))
                            .append("\" aria-label=\"Back to footnote reference ")
                            .append(occurrence).append("\">↩</a>");
                }
                html.append("</span></li>");
            }
            return html.append("</ol></section>").toString();
        }
    }

    private static final class Reference {
        private final String key;
        private final int number;
        private final int occurrence;
        private final int tokenIndex;

        private Reference(String key, int number, int occurrence, int tokenIndex) {
            this.key = key;
            this.number = number;
            this.occurrence = occurrence;
            this.tokenIndex = tokenIndex;
        }
    }
}

File: macos/Sources/LumaMDCore/FootnoteProcessor.swift
import Foundation

internal enum FootnoteProcessor {
    private static let token = "\u{0005}"

    static func extract(_ source: String) -> Result {
        var lines = source.components(separatedBy: "\n")
        let definitions = collectDefinitions(lines)
        var numbers: [String: Int] = [:]
        var occurrences: [String: Int] = [:]
        var references: [Reference] = []
        var inFence = false

        for index in lines.indices {
            let trimmed = lines[index].trimmingCharacters(in: .whitespacesAndNewlines)
            if isFence(trimmed) {
                inFence.toggle()
                continue
            }
            if !inFence {
                if parseDefinition(lines[index]) != nil {
                    lines[index] = ""
                } else {
                    lines[index] = replaceReferences(
                        in: lines[index],
                        definitions: definitions,
                        numbers: &numbers,
                        occurrences: &occurrences,
                        references: &references
                    )
                }
            }
        }

        return Result(
            source: lines.joined(separator: "\n"),
            definitions: definitions,
            numbers: numbers,
            occurrences: occurrences,
            references: references
        )
    }

    private static func collectDefinitions(_ lines: [String]) -> [String: String] {
        var definitions: [String: String] = [:]
        var inFence = false
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if isFence(trimmed) {
                inFence.toggle()
                continue
            }
            if inFence {
                continue
            }
            if let definition = parseDefinition(line), definitions[definition.key] == nil {
                definitions[definition.key] = definition.value
            }
        }
        return definitions
    }

    private static func replaceReferences(
        in line: String,
        definitions: [String: String],
        numbers: inout [String: Int],
        occurrences: inout [String: Int],
        references: inout [Reference]
    ) -> String {
        let characters = Array(line)
        var output = ""
        var inCode = false
        var cursor = 0
        while cursor < characters.count {
            let current = characters[cursor]
            if current == "`" {
                inCode.toggle()
                output.append(current)
                cursor += 1
                continue
            }
            if !inCode && current == "[" && cursor + 3 < characters.count
                    && characters[cursor + 1] == "^" {
                var end = cursor + 2
                while end < characters.count && characters[end] != "]" {
                    end += 1
                }
                if end > cursor + 2 && end < characters.count {
                    let key = String(characters[(cursor + 2)..<end])
                    if definitions[key] != nil {
                        let number: Int
                        if let existing = numbers[key] {
                            number = existing
                        } else {
                            number = numbers.count + 1
                            numbers[key] = number
                        }
                        let occurrence = (occurrences[key] ?? 0) + 1
                        occurrences[key] = occurrence
                        let reference = Reference(
                            key: key,
                            number: number,
                            occurrence: occurrence,
                            tokenIndex: references.count
                        )
                        references.append(reference)
                        output += token + String(reference.tokenIndex) + token
                        cursor = end + 1
                        continue
                    }
                }
            }
            output.append(current)
            cursor += 1
        }
        return output
    }

    private static func isFence(_ line: String) -> Bool {
        line.hasPrefix("```") || line.hasPrefix("~~~")
    }

    private static func parseDefinition(_ line: String) -> (key: String, value: String)? {
        let characters = Array(line)
        var cursor = 0
        var leadingWhitespace = 0
        while cursor < characters.count && characters[cursor].isWhitespace {
            leadingWhitespace += 1
            cursor += 1
        }
        guard leadingWhitespace <= 3,
              cursor + 2 < characters.count,
              characters[cursor] == "[",
              characters[cursor + 1] == "^" else {
            return nil
        }
        cursor += 2
        let keyStart = cursor
        while cursor < characters.count && characters[cursor] != "]" {
            cursor += 1
        }
        guard cursor > keyStart, cursor < characters.count else {
            return nil
        }
        let key = String(characters[keyStart..<cursor])
        cursor += 1
        while cursor < characters.count && characters[cursor].isWhitespace {
            cursor += 1
        }
        guard cursor < characters.count, characters[cursor] == ":" else {
            return nil
        }
        cursor += 1
        while cursor < characters.count && characters[cursor].isWhitespace {
            cursor += 1
        }
        return (key, String(characters[cursor..<characters.count]))
    }

    private static func safeID(_ key: String) -> String {
        var value = ""
        for character in key.lowercased() {
            if character.isLetter || character.isNumber || character == "_" || character == "-" {
                value.append(character)
            } else {
                value.append("-")
            }
        }
        while value.first == "-" {
            value.removeFirst()
        }
        while value.last == "-" {
            value.removeLast()
        }
        return value.isEmpty ? "note" : value
    }

    internal struct Result {
        let source: String
        let definitions: [String: String]
        let numbers: [String: Int]
        let occurrences: [String: Int]
        private let references: [Reference]

        fileprivate init(
            source: String,
            definitions: [String: String],
            numbers: [String: Int],
            occurrences: [String: Int],
            references: [Reference]
        ) {
            self.source = source
            self.definitions = definitions
            self.numbers = numbers
            self.occurrences = occurrences
            self.references = references
        }

        func finish(
            _ documentHTML: String,
            renderDefinition: ((String) -> String)? = nil
        ) -> String {
            var html = documentHTML
            for reference in references {
                html = html.replacingOccurrences(
                    of: FootnoteProcessor.token + String(reference.tokenIndex) + FootnoteProcessor.token,
                    with: referenceHTML(reference)
                )
            }
            return html + sectionHTML(renderDefinition: renderDefinition)
        }

        private func referenceHTML(_ reference: Reference) -> String {
            let id = FootnoteProcessor.safeID(reference.key)
            let referenceID = "fnref-" + id
                + (reference.occurrence > 1 ? "-" + String(reference.occurrence) : "")
            return "<sup class=\"footnote-ref\"><a id=\""
                + FootnoteProcessor.escape(referenceID)
                + "\" href=\"#"
                + FootnoteProcessor.escape("fn-" + id)
                + "\" aria-label=\"Footnote "
                + String(reference.number)
                + "\">"
                + String(reference.number)
                + "</a></sup>"
        }

        private func sectionHTML(renderDefinition: ((String) -> String)?) -> String {
            guard !numbers.isEmpty else { return "" }
            var html = "<section class=\"footnotes\" aria-label=\"Footnotes\"><hr><ol>"
            for (key, _) in numbersInReferenceOrder() {
                let id = FootnoteProcessor.safeID(key)
                html += "<li id=\"fn-" + FootnoteProcessor.escape(id) + "\"><span class=\"footnote-content\">"
                let definition = definitions[key] ?? ""
                html += renderDefinition?(definition) ?? FootnoteProcessor.renderInline(definition)
                html += "</span><span class=\"footnote-backlinks\">"
                let count = occurrences[key] ?? 0
                if count > 0 {
                    for occurrence in 1...count {
                        let referenceID = "fnref-" + id
                            + (occurrence > 1 ? "-" + String(occurrence) : "")
                        html += "<a class=\"footnote-backref\" href=\"#"
                            + FootnoteProcessor.escape(referenceID)
                            + "\" aria-label=\"Back to footnote reference "
                            + String(occurrence)
                            + "\">↩</a>"
                    }
                }
                html += "</span></li>"
            }
            return html + "</ol></section>"
        }

        private func numbersInReferenceOrder() -> [(String, Int)] {
            numbers.sorted { left, right in
                if left.value == right.value {
                    return left.key < right.key
                }
                return left.value < right.value
            }
        }
    }

    fileprivate struct Reference {
        let key: String
        let number: Int
        let occurrence: Int
        let tokenIndex: Int
    }

    private static func renderInline(_ source: String) -> String {
        let escaped = escape(source)
        var output = escaped
        output = replaceDelimited(output, marker: "**", open: "<strong>", close: "</strong>")
        output = replaceDelimited(output, marker: "*", open: "<em>", close: "</em>")
        output = replaceDelimited(output, marker: "`", open: "<code>", close: "</code>")
        return output
    }

    private static func replaceDelimited(
        _ source: String,
        marker: String,
        open: String,
        close: String
    ) -> String {
        var output = ""
        var remainder = source
        while let start = remainder.range(of: marker),
              let end = remainder.range(of: marker, range: start.upperBound..<remainder.endIndex) {
            output += remainder[..<start.lowerBound]
            output += open
            output += remainder[start.upperBound..<end.lowerBound]
            output += close
            remainder = String(remainder[end.upperBound...])
        }
        return output + remainder
    }

    private static func escape(_ value: String) -> String {
        value
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&#39;")
    }
}
